### PR TITLE
[WIP] check Fortran contiguity in valid_audio

### LINKED ISSUE #964: util.valid_audio validates non-Fortran contiguous arrays
#### Description
Raised during CR of #960 which cl0sed #959.
https://github.com/librosa/librosa/pull/960#issuecomment-521862461

Last time we updated `valid_audio` was #616 which cl0sed #612

#### Steps/Code to Reproduce

```python
from librosa import util
import numpy as np

y = np.zeros(1000)
print(util.valid_audio(y))

y_strided = np.zeros(1000)[::2]
print(util.valid_audio(y_strided))

y_clang = np.zeros(1000).reshape(2, 500)
print(util.valid_audio(y_clang, mono=False))

y_fortran = np.asfortranarray(
    np.zeros(1000).reshape(2, 500))
print(util.valid_audio(y_fortran, mono=False))

y_fortran_colon = y_fortran[:]
print(util.valid_audio(y_fortran_colon, mono=False))
```

#### Expected Results
```python
True
False
False
True
True
```

#### Actual Results
```python
True
True
True
True
True
```

#### Versions
```
Darwin-18.6.0-x86_64-i386-64bit
Python 3.7.4 (default, Aug  9 2019, 12:36:10) 
[Clang 4.0.1 (tags/RELEASE_401/final)]
NumPy 1.16.4
SciPy 1.3.1
librosa 0.7.0
```

### PR DESCRIPTION
**Pull request by lostanlen**
#### Reference Issue
Fixes #964

#### What does this implement/fix? Explain your changes.
This really amounts to:

```python
    if not y.flags["F_CONTIGUOUS"]:
        raise ParameterError('Audio buffer is not Fortran-contiguous. '
            'Use numpy.asfortranarray to ensure Fortran contiguity.')
```

Note that i refer to `librosa.stack` in the "See Also" section, but i don't speak of it in the error message, because there can be many possible causes behind `y.flags["F_CONTIGUOUS"]`, so i think it's best to not assume anything and just advise for `np.asfortranarray` even though that might not be most efficient solution. (i would expect that `librosa.util.stack` is faster than `np.stack` followed by `np.asfortranarray`?)

#### Any other comments?
I also made small changes to the docstring to clarify its phrasing.

### CODE REVIEW

**Comment by lostanlen** [comment]
I'm very confused by the failing tests:


```python
    def test_beat():
    
>       y, sr = librosa.load(__EXAMPLE_FILE)
tests/test_beat.py:117: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
librosa/core/audio.py:153: in load
    y = to_mono(y)
librosa/core/audio.py:424: in to_mono
    util.valid_audio(y, mono=False)
```


```python
>       if not y.flags["F_CONTIGUOUS
    "]:
E   KeyError: 'Unknown flag'
librosa/util/utils.py:184: KeyError
```

does anyone have a clue about this?

**Comment by lostanlen** [comment]
Goals: 
* inquire about the unknown flag error
* have to_mono return F-contiguous arrays (as per #964 latest comment)

File: librosa/util/utils.py [hunk 1]
```diff
@@ -225,47 +225,59 @@ def valid_audio(y, mono=True):
...
     elif y.ndim > 2 or y.ndim == 0:
-        raise ParameterError('Audio must have shape (samples,) or (channels, samples). '
+        raise ParameterError('Audio data must have shape (samples,) or (channels, samples). '
                              'Received shape={}'.format(y.shape))
 
     if not np.isfinite(y).all():
         raise ParameterError('Audio buffer is not finite everywhere')
 
+    if not y.flags["F_CONTIGUOUS
"]:
```
Comment (bmcfee) [inline librosa/util/utils.py]: Aha!  The problem is that you have a unicode garbage character hidden at the end of this string.

```"F_CONTIGUOUS"```

**Comment by lostanlen** [comment]
I fixed the spooky invisible Unicode character bug 👻 
The tests are still failing, but now for the right reason.
Will update `to_mono` to ensure Fortran contiguity now

**Comment by lostanlen** [comment]
I'm seeing a test failure in `effects.remix`. this is because we validate before converting to mono. 
I will enforce contiguity before validating.

**Comment by lostanlen** [comment]
ok i'm validating the output!

**Comment by bmcfee** [comment]
LGTM!  Should we merge?

File: librosa/core/audio.py [hunk 2]
```diff
@@ -638,6 +641,10 @@ def get_duration(y=None, sr=22050, S=None, n_fft=2048, hop_length=512,
             n_samples = n_samples - 2 * int(n_fft / 2)
 
     else:
+        # Ensure Fortran contiguity.
+        if not np.isfortran(y):
+            y = np.asfortranarray(y)
```
Comment (bmcfee) [inline librosa/core/audio.py]: This is equivalent to `y = np.asfortranarray(y)`, which returns the input unmodified if it's already F-contig.

```python
In [2]: x = np.ones(5)                                                                        

In [3]: x.flags['F_CONTIGUOUS']                                                               
Out[3]: True

In [4]: y = np.asfortranarray(x)                                                              

In [5]: y is x                                                                                
Out[5]: True
```

File: librosa/util/utils.py [hunk 3]
```diff
@@ -225,47 +225,59 @@ def valid_audio(y, mono=True):
     Raises
     ------
     ParameterError
-        If `y` fails to meet the following criteria:
-            - `type(y)` is `np.ndarray`
-            - `y.dtype` is floating-point
+        In either of these cases:
```
Comment (bmcfee) [inline librosa/util/utils.py]: "either" => "any"

Comment (lostanlen) [reply librosa/util/utils.py]: fixed in ccf3aa26

File: librosa/core/audio.py (hunk 2 above)
Comment (lostanlen) [reply librosa/core/audio.py]: fixed in d4decceb

**Comment by lostanlen** [comment]
Thank you for the CR. I made the two required changes.

**Comment by bmcfee** [comment]
Lgtm! Merge at your leisure.

**Comment by bmcfee** [comment]
Anything else to add here?

**Comment by lostanlen** [comment]
no, now it's good to go.

**Comment by Alan-Zhao-908** [comment]
Hey guys sorry to ask this again but I'm getting the same error when I run the following code:

`def save_data_to_array(path=DATA_PATH, max_len=11, n_mfcc=20):
    labels, _, _ = get_labels(path)

    for label in labels:
        # Init mfcc vectors
        mfcc_vectors = []

        wavfiles = [path + label + '/' + wavfile for wavfile in os.listdir(path + '/' + label)]
        for wavfile in tqdm(wavfiles, "Saving vectors of label - '{}'".format(label)):
            mfcc = wav2mfcc(wavfile, max_len=max_len, n_mfcc=n_mfcc)
            mfcc_vectors.append(mfcc)
        np.save(label + '.npy', mfcc_vectors)

import keras
from keras.models import Sequential
from keras.layers import Dense, Dropout, Flatten, Conv2D, MaxPooling2D, LSTM
from keras.utils import to_categorical
import wandb
from wandb.keras import WandbCallback
import matplotlib.pyplot as plt

wandb.init()
config = wandb.config

config.max_len = 11
config.buckets = 20

# Save data to array file first
save_data_to_array(max_len=config.max_len, n_mfcc=config.buckets)
`

```ParameterError                            Traceback (most recent call last)
<ipython-input-100-d8cda3907224> in <module>
----> 1 save_data_to_array(max_len=config.max_len, n_mfcc=config.buckets)

<ipython-input-99-8a7c3bf16e6d> in save_data_to_array(path, max_len, n_mfcc)
      8         wavfiles = [path + label + '/' + wavfile for wavfile in os.listdir(path + '/' + label)]
      9         for wavfile in tqdm(wavfiles, "Saving vectors of label - '{}'".format(label)):
---> 10             mfcc = wav2mfcc(wavfile, max_len=max_len, n_mfcc=n_mfcc)
     11             mfcc_vectors.append(mfcc)
     12         np.save(label + '.npy', mfcc_vectors)

~/Documents/coding/Python_Notebooks/Neural_Network_Tutorials/preprocess.py in wav2mfcc(file_path, n_mfcc, max_len)
     22     wave, sr = librosa.load(file_path, mono=True, sr=None)
     23     wave = wave[::3]
---> 24     mfcc = librosa.feature.mfcc(wave, sr=16000, n_mfcc=n_mfcc)
     25 
     26     # If maximum length exceeds mfcc lengths then pad the remaining ones

/anaconda3/lib/python3.7/site-packages/librosa/feature/spectral.py in mfcc(y, sr, S, n_mfcc, dct_type, norm, lifter, **kwargs)
   1690 
   1691     if S is None:
-> 1692         S = power_to_db(melspectrogram(y=y, sr=sr, **kwargs))
   1693 
   1694     M = scipy.fftpack.dct(S, axis=0, type=dct_type, norm=norm)[:n_mfcc]

/anaconda3/lib/python3.7/site-packages/librosa/feature/spectral.py in melspectrogram(y, sr, S, n_fft, hop_length, win_length, window, center, pad_mode, power, **kwargs)
   1815     S, n_fft = _spectrogram(y=y, S=S, n_fft=n_fft, hop_length=hop_length, power=power,
   1816                             win_length=win_length, window=window, center=center,
-> 1817                             pad_mode=pad_mode)
   1818 
   1819     # Build a Mel filter

/anaconda3/lib/python3.7/site-packages/librosa/core/spectrum.py in _spectrogram(y, S, n_fft, hop_length, power, win_length, window, center, pad_mode)
   2526         S = np.abs(stft(y, n_fft=n_fft, hop_length=hop_length,
   2527                         win_length=win_length, center=center,
-> 2528                         window=window, pad_mode=pad_mode))**power
   2529 
   2530     return S, n_fft

/anaconda3/lib/python3.7/site-packages/librosa/core/spectrum.py in stft(y, n_fft, hop_length, win_length, window, center, dtype, pad_mode)
    213 
    214     # Check audio is valid
--> 215     util.valid_audio(y)
    216 
    217     # Pad the time series so that frames are centered

/anaconda3/lib/python3.7/site-packages/librosa/util/utils.py in valid_audio(y, mono)
    276 
    277     if not y.flags["F_CONTIGUOUS"]:
--> 278         raise ParameterError('Audio buffer is not Fortran-contiguous. '
    279             'Use numpy.asfortranarray to ensure Fortran contiguity.')
    280 

ParameterError: Audio buffer is not Fortran-contiguous. Use numpy.asfortranarray to ensure Fortran contiguity.```

Any thoughts on how to resolve this?

**Pull request merged**